### ISSUE TITLE
Warn about unneeded parentheses

### ISSUE THREAD

**Issue by olivierperes** [mgechev/revive#159]
**Is your feature request related to a problem? Please describe.**
This passes with no warnings:
```go
package main

import "fmt"

func main() {
	x := ((((1+2))))
	fmt.Println(x)
}
```
**Describe the solution you'd like**
A warning about useless parentheses.

**Comment by chavacava** [comment]
Hi @olivierperes, do you expect warnings for which of the following expressions:

```go
y := ((1)+(2)) 
z := (1 * 2) + 3
a := (1 + 2) + 3
b := (1 * 2) * 3
c := (1 * 2)
```

**Comment by olivierperes** [comment]
I was only thinking about cases in which they are pathologically useless, not even potentially improving readability, so only your first and last examples.

**Comment by mgechev** [comment]
That's a fun problem to solve!

A potential algorithm could be:
- Parse the expression in an AST where the operation priority and the parentheses determine the position in the tree
- Serialize the tree to a string adding the minimum amount of parentheses, depending on the operations priority
- Compare the serialized expression with the original expression
- Warn in case the expressions do not match

**Comment by olivierperes** [comment]
This is a possible approach, but I fear it could be too restrictive. In a complex mathematical or logical expression, it might be good to keep a few unnecessary parentheses around, just so humans are convinced that the author got the priorities right. Of course, one could argue that such expressions should not be used in the first place, but broken down into simpler expressions.

**Comment by mgechev** [comment]
How would revive distinguish between extra parentheses used for readability versus redundant ones that are in the expression because negligence?

**Comment by chavacava** [comment]
An example of such an expression is

(1 * 2) + 3

Or more abstractly

( exp1 * exp2) + exp3

 Should we warn or not? It is a matter of "taste"... but algorithms do not have taste

**Comment by olivierperes** [comment]
I think it should be possible to detect repeated parentheses, as in `((expr))`. Another simple case is when the expression directly after `=` or `:=` is between parentheses, as in `x := (expr)` or `x = (expr)`. I suppose the case of multiple assignment might be trickier. I don’t have a strong opinion as to whether anything deeper should be attempted, either a strict approach or a more relaxed one would suit me.

**Comment by mgechev** [comment]
Looking at the problem and the proposed solution https://github.com/mgechev/revive/issues/159#issuecomment-504516839, this sounds like a better fit for a formatter, not a linter.

**Comment by olivierperes** [comment]
Hmm, I don’t know. Does gofmt suppress (non-whitespace) characters? It certainly moves them around, but more than that? I could always open an issue in that project and see what comes out of it.
